Web UI disconnects

a-palchikov opened the pull request:
This PR fixes an issue with web UI sessions sporadically disconnecting.

In the scenario with multiple proxy instances running behind a load balancer (e.g. a kubernetes service)
the web session contexts (i.e. the scope of a single logged in user) were stored in memory in the corresponding proxy process upon creation. This also included open terminal handles. If the web session renewal request came to another proxy as the one that previously created the session, the session was renewed out-of-band. Once the internal cache clean up loop running inside of each proxy instance detected the expired session, it could release the resources associated with the previous session and forcefully close the terminals.
Additionally, a client holding to the old bearer token (until is has gotten the update) after the session has been renewed, might access the API using the expired token leading to the access denied error that would trigger the logout.

Now, the following counter measures were introduced:
  1. Renewing a session does not release any resources that were created when the initial session was created. This is implemented by persisting the resources on the user instead of the session level.
  1. Bearer tokens are not explicitly removed upon session renewal but are instead let to linger and expire naturally minimizing the chances of a "late" client

A new web token resource type is introduced and both web sessions and web tokens are made available in the cache layer to make validation logic cheaper.
As web sessions and web tokens (used as bearer tokens in this context) are two separate resources, independent lifecycle management is possible.
To simplify the implementation, web sessions are still constructed together with bearer tokens - e.g. a web session value
is used as a container for the initial bearer token. Once propagated to a client, bearer tokens are queried separately from web sessions and have independent expiration times.
To be able to cache the values of the new types, I chose a simpler backend storage story: instead of keeping the sessions under the user:
```
  /web/users/<user>/sessions/<id>
```
the sessions were moved up into its own namespace:
```
  /web/sessions/<id>
```
The same is with the new web tokens resource:
```
  /web/tokens/<id>
```
althoguh both resources are still bound to a user.
Apart from simpler implementation there was an issue I saw with the caching layer that does not have ordering guarantees in the way it handles refreshes - e.g. it seemed possible that had I kept the sessions/tokens under the user, the cache could invalidate the sessions/tokens before doing so for users and the users update would remove all sessions/tokens data due to the nesting in the backend - I'm open to suggestions here though.

No changes to base behavior were introduced:
 1. Bearer tokens are created with a fixed TTL (auth.BearerTokenTTL)
 1. Sessions have a top-level expiration but are capped at bearer token's TTL and are periodically renewed

A note on backwards-compatibility: since the data was moved to an alternative path in the backend, provisions were made to still read the data in the old path if not available in the new path.

The prominent changes:
  [lib/web/sessions.go](https://github.com/gravitational/teleport/compare/dmitri/web-ui-disconnects#diff-7e877d49db0989e2f62f90da78aa9a7ce5b3c665ea94f63a74d97c2ba34623ee)
  [lib/web/apiserver.go](https://github.com/gravitational/teleport/compare/dmitri/web-ui-disconnects#diff-5a12dd7747d28af8fe5043638cc11f31a3314b2a23f3e23fe0bcdc952a805fa6)
  [lib/cache/cache.go](https://github.com/gravitational/teleport/compare/dmitri/web-ui-disconnects#diff-3e237f973a30f7092dc18a706ecd95d8a7e1d99c5c50626b8bb0e7ac6cb7efd1)
  [lib/cache/collections.go](https://github.com/gravitational/teleport/compare/dmitri/web-ui-disconnects#diff-cb7927b8cb36bb63bc941389c91c41c50c69d81f892361c9f8ce939092e105fb)


Fixes https://github.com/gravitational/teleport/issues/5265

a-palchikov committed:
wip: web sessions caching layer

a-palchikov committed:
Split web session management into two interfaces and implement them separately for clear separation

a-palchikov committed:
Split session management into New/Validate to make it aparent where the
sessions are created and where existing sessions are managed.
Remove ttlmap in favor of a simple map and handle expirations
explicitly.
Add web session management to gRPC server for the cache.

a-palchikov committed:
Reintroduce web sessions APIs under a getter interface.

a-palchikov committed:
Add SubKind to WatchKind for gRPC and add conversions from/to protobuf.
Fix web sessions unit tests.

a-palchikov committed:
lib/web: create/insert session context in ValidateSession if the session
has not yet been added to session cache.
lib/cache: add event filter for web session in auth cache.
lib/auth: propagate web session subkind in gRPC event.

a-palchikov committed:
Add implicit migrations for legacy web session key path for queries.

a-palchikov committed:
wip: adding web token resource

a-palchikov committed:
Integrate web token in lib/web

a-palchikov committed:
Add a bearer token when upserting a web session

a-palchikov committed:
Fix tests. Use fake clock wherever possible.

a-palchikov committed:
Converge session cache handling in lib/web

a-palchikov committed:
Clean up and add doc comments where necessary

a-palchikov committed:
Use correct form of sessions/tokens controller for ServerWithRoles. Use fake time in web tests

a-palchikov committed:
Converge the web sessions/tokens handling in lib/auth to match
the old behavior w.r.t access checking (e.g. implicit handling
of the local user identity).

a-palchikov committed:
Use fake clock consistently in units tests.

a-palchikov committed:
Use cached reads and waiters only when necessary. Query sessions/tokens using best-effort - first looking in the cache and falling back to a proxy client

a-palchikov committed:
Properly propagate events about deletes for values with subkind.

a-palchikov committed:
Update to retrofit changes after recent teleport API refactorings

a-palchikov committed:
fix lint warnings

a-palchikov commented on api/types/session.go:
```diff
@@ -84,6 +82,8 @@ func NewWebSession(name string, kind string, subkind string, spec WebSessionSpec
 		},
 		Spec: spec,
 	}
+	session.Metadata.SetExpiry(spec.Expires)
```
I believe I misinterpreted `Metadata.Expires` and `Session.Spec.Expires`. Metadata.Expires is meant as a resource expiry in backend context while `Spec.Expires` declares the general expiration and these might differ - I'll update.

a-palchikov replied:
No, I believe this is a correct assumption. I have not found a counter argument in the current code base and it seems that other resources (i.e. semaphores) default to the same behavior.

a-palchikov committed:
Update comment on removing legacy code to move the deadline to 7.x

a-palchikov committed:
Do not close the resources on the session when it expires - this beats
the purpose of this PR.
Also avoid a race between closing the cached clients and an existing
reference to the session by letting the session linger for longer before
removing it.

r0mant commented on lib/services/identity.go:
```diff
@@ -246,6 +248,54 @@ type AppSession interface {
+// WebTokenInterface defines interface for managing web tokens
+type WebTokenInterface interface {
+	// Get returns a token specified by the request.
+	Get(ctx context.Context, req types.GetWebTokenRequest) (types.WebToken, error)
```
IMHO the generic names of the methods in this interface and `WebSessionInterface` below make it unfriendly to embedding since the names will clash. All other interfaces I think call them like `GetWebToken`, `DeleteAllWebTokens`, etc. Have you considered doing the same?

I see you have respective `XXXGetter` interfaces likely for this reason, maybe they also won't be needed if you just named the methods with "full" names? Unless I'm missing something.

a-palchikov replied:
I still have a case when I need to support a couple of legacy APIs and I honestly did not want to come up with yet another prefixed or suffixed alternatives which required a different set of parameters. Also, I feel strongly against the flat namespace and I liked the direction that the kubernetes client took with its [Clientset](https://github.com/kubernetes/client-go/blob/c7a570ddfedd593495d663befa0f94bd127eac66/kubernetes/clientset.go#L70) and, for example, [CoreV1](https://github.com/kubernetes/client-go/blob/c7a570ddfedd593495d663befa0f94bd127eac66/kubernetes/typed/core/v1/core_client.go#L27) hence I made a small step towards the same thing here - using wrapper values with a well-defined set of APIs for all resources and differentiating them at the API entry with a corresponding getter.

The original layout is also unfriendly to versioning and this will eventually surface as an issue we'll need to solve.

r0mant commented on lib/services/local/session.go:
```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
 }
+
+// GetWebSession returns a web session state for the given user and session id
+func (s *IdentityService) GetWebSession(user, sessionID string) (services.WebSession, error) {
```
Nit: Since you're refactoring all these methods anyway, might be a good excuse to enhance them to accept contexts? Feel free to disregard if that's too much work / out of scope.

a-palchikov replied:
I'd consider this out of scope for now due to the amount of additional changes. Will revisit in the future.

r0mant commented on lib/services/local/session.go:
```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	session, err := services.GetWebSessionMarshaler().UnmarshalWebSession(item.Value, services.SkipValidation())
```
I'd just start importing from `types` instead of `services` in the new code to simplify future refactoring.

r0mant commented on lib/services/local/session.go:
```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+}
+
+func webSessionKey(sessionID string) (key []byte) {
+	return backend.Key(webPrefix, sessionsPrefix, sessionID)
```
So with this change to store all sessions under the same key now, would it still be possible, for example, to implement features like "remove all sessions for a particular user" if we wanted? It'd just require iterating over all sessions?

a-palchikov replied:
Yes, the downside is that it requires to iterate over all values in order to remove session only for a specific user. As the web/app sessions (and tokens) are explicitly expired, this is not such a big deal but might need to be fixed nonetheless at some point.

r0mant commented on lib/services/local/session.go:
```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+	item := backend.Item{
+		Key:     webSessionKey(session.GetName()),
+		Value:   value,
+		Expires: backend.EarliestExpiry(session.GetBearerTokenExpiryTime(), sessionMetadata.Expiry()),
```
Nit: Doesn't resource already have `GetExpiryTime()` method, so you prob. don't need to extract metadata first?

a-palchikov replied:
I kept the original behavior as much as possible where the `Metadata.Expires` is used as a resource expiration timestamp both ways. Although in case of web sessions the Spec.Expires is sort of redundant b/c of that. I'm on the fence but would probably be keen on keeping the redundancy.

r0mant commented on lib/web/sessions.go:
```diff
@@ -342,19 +393,26 @@ func newSessionCache(config *sessionCache) (*sessionCache, error) {
+	// contexts maps user to an active user web session.
+	// These are used to maintain session state that would otherwise
+	// not survive renewal of the session
+	contexts map[string]*sessionContext
```
Sorry just looking at this is a little confusing - why are there two "session context" structs, one exported and one not, and what's the difference b/w them? And the fact that one field is called "sessions" and another one "contexts" is also a bit confusing.

a-palchikov replied:
It is - I'll update to reflect the actual purpose of the unexpected context.

r0mant commented on lib/web/sessions.go:
```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
+func (s *sessionCache) insertContext(user string, ctx *SessionContext) (exists bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	id := user + ctx.session.GetName()
```
I'd consider making a function, `makeKey` or something, that implements this concatenation logic since it's used in a few places.

r0mant commented on lib/web/sessions.go:
```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
-		if trace.IsAlreadyExists(err) {
-			defer c.Close()
-			return out, nil
+func (c *sessionContext) Close() error {
```
Nit: Add godoc.

r0mant commented on lib/web/sessions.go:
```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
+	closers := c.transferClosers()
+	for _, closer := range closers {
+		c.log.Debugf("Closing %v.", closer)
+		closer.Close()
```
Doesn't this method also return an error?

r0mant commented on lib/web/sessions.go:
```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
-		HttpOnly: true,
-		Secure:   true,
-	})
+func waitForResource(ctx context.Context, watcher services.Watcher, m eventMatcher, clock clockwork.Clock) (services.Resource, error) {
```
This feels like it doesn't really belong in the web package TBH, maybe move this somewhere to services closer to the `Watcher`, along with `eventMatcher`?

a-palchikov commented on api/types/types.proto:
```diff
@@ -1351,3 +1351,66 @@ message KubernetesCluster {
+}
+
+// WebTokenV1 describes a web token
+message WebTokenV1 {
```
Ok, was wondering myself. Will update.

a-palchikov commented on api/client/proto/authservice.proto:
```diff
@@ -477,6 +507,24 @@ service AuthService {
     rpc DeleteAllAppSessions(google.protobuf.Empty) returns (google.protobuf.Empty);
 
+    // GetWebSession gets a web session.
+    rpc GetWebSession(types.GetWebSessionRequest) returns (GetWebSessionResponse);
```
The reasoning is that the request is not gRPC-specific whilte the response is.

a-palchikov replied:
Moved them to the proto package.

a-palchikov commented on api/defaults/defaults.go:
```diff
@@ -42,9 +42,12 @@ const (
+	// MaxCertDuration limits maximum duration of validity of issued certificate
 	MaxCertDuration = 30 * time.Hour
 
+	// CertDuration is a default certificate duration.
```
Hmm, good question - I'll see if I can find out from history.

a-palchikov commented on api/types/events.go:
```diff
@@ -18,10 +18,19 @@ package types
 
+// String returns text description of this event
+func (r Event) String() string {
+	if r.Type == OpDelete {
```
It's actually about a slightly different thing - the events for a deleted resource include only the header in `Resource` and formatting it with `fmt.SprintX` panics.

a-palchikov commented on api/types/session.go:
```diff
@@ -412,3 +408,274 @@ func GetWebSessionMarshaler() WebSessionMarshaler {
+
+// CheckAndSetDefaults validates this token value and sets defaults
+func (r *WebTokenV1) CheckAndSetDefaults() error {
+	return r.Metadata.CheckAndSetDefaults()
```
Yes - will update.

a-palchikov commented on lib/cache/cache.go:
```diff
@@ -57,7 +58,9 @@ func ForAuth(cfg Config) Config {
 		{Kind: services.KindAppServer},
-		{Kind: services.KindWebSession},
+		{Kind: services.KindWebSession, SubKind: services.KindAppSession},
+		{Kind: services.KindWebSession, SubKind: services.KindWebSession},
```
Yes, it's a bit complicated.
The app sessions already used the subkind but enjoyed the status of being the only web session in the caching layer hence the caching layer did not originally require any modifications. Also, I see a legit need to support resources with a subkind in cache hence I went forward with extending it to support that.

a-palchikov committed:
Address review comments

a-palchikov committed:
Move web session/token request structs to the api client proto package

a-palchikov commented on lib/auth/clt.go:
```diff
@@ -2117,6 +2118,15 @@ func (c *Client) CreateAuditStream(ctx context.Context, sid session.ID) (events.
+	services.AppSession
+}
+
+// WebUIService implements features used by Web UI clients
```
Removed one of them - was no longer necessary to split.

r0mant commented on lib/services/local/events.go:
```diff
@@ -832,6 +887,72 @@ func resourceHeader(event backend.Event, kind, version string, offset int) (serv
+		select {
+		case event := <-watcher.Events():
+			res, err := m.Match(event)
+			if err == nil {
```
What if `err != nil`?

a-palchikov replied:
I'll log the error in that case.

r0mant commented on lib/web/apps.go:
```diff
@@ -174,18 +175,20 @@ func (h *Handler) createAppSession(w http.ResponseWriter, r *http.Request, p htt
-
+func (h *Handler) waitForAppSession(ctx context.Context, sessionID string) error {
+	_, err := h.cfg.AccessPoint.GetAppSession(ctx, services.GetAppSessionRequest{SessionID: sessionID})
+	if err == nil {
```
Handle error other than not found?

a-palchikov replied:
I've added the log message similar to web session wait.

r0mant commented on lib/web/apps.go:
```diff
@@ -194,48 +197,22 @@ func (h *Handler) waitForSession(ctx context.Context, sessionID string) error {
-			return trace.BadParameter("expected init event, got %v instead", event.Type)
+	matchEvent := func(event services.Event) (services.Resource, error) {
+		if event.Type == backend.OpPut &&
+			event.Resource.GetKind() == services.KindWebSession &&
```
Should this also check the subkind?

r0mant commented on lib/web/sessions.go:
```diff
@@ -282,55 +280,107 @@ func (c *SessionContext) GetAgent() (agent.Agent, *ssh.Certificate, error) {
+	// This is a cheaper way to avoid race without introducing object
+	// reference counters.
+	// TODO(dmitri): this leaks the implementation detail about the session TTL
+	return c.parent.clock.Since(expiry) > 2*auth.BearerTokenTTL
```
Maybe I'm reading this incorrectly, but doesn't this basically extend the session TTL by 3 times? It allows the session to linger for 20 minutes after expiration. Is this not going to be a security concern?

a-palchikov replied:
This test is only for the proxy's internal memory cache - it does not affect expiration of the actual sessions in the backend. As given in the comment, it is the attempt to provide cheap protection against a client accessing served SessionContexts (which have cluster client caches) while the context is removed in parallel. The hope is that SessionContext instance should goes out of scope sooner than the double of the TTL.

r0mant commented on lib/web/sessions.go:
```diff
@@ -497,148 +555,302 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
+	defer watcher.Close()
+	matchEvent := func(event services.Event) (services.Resource, error) {
+		if event.Type == backend.OpPut &&
+			event.Resource.GetKind() == services.KindWebSession &&
```
Check subkind or that would be redundant?

a-palchikov replied:
I'll add the subkind although the IDs should be unique between (even types of) sessions.

r0mant reviewed (approved):
Code lgtm, I'd get a 2nd pair of eyes on the session/token management changes.

a-palchikov commented on api/client/proto/proto.go:
```diff
@@ -0,0 +1,85 @@
+// Check validates the request.
+func (r *GetWebSessionRequest) Check() error {
+	if r.User == "" {
+		return trace.BadParameter("user name missing")
```
Wholeheartedly agree for the consistency in these messages. Just need to settle on the form.

a-palchikov committed:
Address review comments

awly commented on api/types/session.go:
```diff
@@ -413,3 +411,245 @@ func GetWebSessionMarshaler() WebSessionMarshaler {
+		},
+		Spec: spec,
+	}
+	token.Metadata.SetExpiry(spec.Expires)
```
nit: can you set Expiry in the literal above?

awly commented on api/types/types.proto:
```diff
@@ -1415,3 +1415,37 @@ message KubernetesCluster {
+    // Token specifies the token's value.
+    string Token = 2 [ (gogoproto.jsontag) = "token" ];
+    // Expires specifies the token's expiration time.
+    google.protobuf.Timestamp Expires = 3 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "expires"
+    ];
```
there's already an `Expires` field in `Metadata`, remove this

awly commented on lib/services/local/events.go:
```diff
@@ -832,6 +887,75 @@ func resourceHeader(event backend.Event, kind, version string, offset int) (serv
+
+// WaitForEvent waits for the event matched by the specified event matcher in the given watcher.
+func WaitForEvent(ctx context.Context, watcher services.Watcher, m EventMatcher, clock clockwork.Clock) (services.Resource, error) {
+	tick := clock.NewTicker(defaults.WebHeadersTimeout)
```
nit: you probably want `NewTimer`

a-palchikov replied:
Unfortunately  there's no timer support in clockwork.

awly commented on lib/services/local/session.go:
```diff
@@ -95,3 +98,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
 }
+
+// GetWebSession returns a web session state for the given user and session id
+func (s *IdentityService) GetWebSession(user, sessionID string) (types.WebSession, error) {
```
here and below, can you take the context as an argument?

awly commented on lib/services/local/session.go:
```diff
@@ -95,3 +98,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+
+// UpsertWebSession updates or inserts a web session for the user and the session ID
+func (s *IdentityService) UpsertWebSession(user, sessionID string, session types.WebSession) error {
+	session.SetUser(user)
+	session.SetName(sessionID)
```
shouldn't the caller populate the session fields?

a-palchikov replied:
This is from my attempt to keep the old APIs for whatever reason. I now believe I can remove these as well.

awly commented on lib/services/local/session.go:
```diff
@@ -95,3 +98,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+
+// GetWebSession returns a web session state for the given user and session id
+func (s *IdentityService) GetWebSession(user, sessionID string) (types.WebSession, error) {
+	return s.WebSessions().Get(context.TODO(), proto.GetWebSessionRequest{
```
nit: trace.Wrap

awly commented on lib/services/local/session.go:
```diff
@@ -95,3 +98,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+func (s *IdentityService) UpsertWebSession(user, sessionID string, session types.WebSession) error {
+	session.SetUser(user)
+	session.SetName(sessionID)
+	return s.WebSessions().Upsert(context.TODO(), session)
```
nit: trace.Wrap

awly commented on lib/web/apiserver_test.go:
```diff
@@ -606,8 +607,15 @@ func (s *WebSuite) TestPasswordChange(c *C) {
 }
 
 func (s *WebSuite) TestWebSessionsRenew(c *C) {
```
optional: rewrite this test following https://github.com/gravitational/teleport/blob/master/rfd/0001-testing-guidelines.md if you have the time

a-palchikov replied:
Done although it turned into a bit more changes than anticipated: https://github.com/gravitational/teleport/pull/5276/commits/308177f1d15865a52ccec7b6452c3171cae05228 and https://github.com/gravitational/teleport/pull/5276/commits/4ba1f5396aaf467da1b1ad39f96cf4d633aed366

awly reviewed (commented):
ugh, github butchered some of these diffs...

a-palchikov commented on api/client/proto/proto.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/gravitational/trace"
+)
+
+// FromWatchKind converts the watch kind value between internal
+// and the protobuf format
+func FromWatchKind(wk types.WatchKind) WatchKind {
+	return WatchKind{
+		Name:        wk.Name,
+		Kind:        wk.Kind,
+		SubKind:     wk.SubKind,
+		LoadSecrets: wk.LoadSecrets,
+		Filter:      wk.Filter,
+	}
+}
+
+// ToWatchKind converts the watch kind value between the protobuf
+// and the internal format
+func ToWatchKind(wk WatchKind) types.WatchKind {
+	return types.WatchKind{
+		Name:        wk.Name,
+		Kind:        wk.Kind,
+		SubKind:     wk.SubKind,
+		LoadSecrets: wk.LoadSecrets,
+		Filter:      wk.Filter,
+	}
+}
```
This is good question. I've had a similar concern with @Joerger api refactoring [PR](https://github.com/gravitational/teleport/pull/5384#discussion_r564455486) - the stand there is that the api client needs to manage a separate set of types unrelated to protobuf to avoid leaking implementation details about internal use of gRPC which I have trouble justifying.
If, on the hand, the request types were to be removed from the protobuf side, it would look equally strange. So some consensus is required.

Joerger replied:
The way we deduplicate this currently is to define the type in `api/types/types.proto`. You can then add methods in `types/sessions.go`, and import the types into `authservice.proto`. You can look at types like `types.CertAuthorityV2` for an example of that. 

But the problem here is that WatchKind is defined in `authservice.proto`. It should probably be deprecated and moved to `types.proto` to solve this issue. This is also an issue with `Events` and `Duration` which should likewise be refactored eventually.

a-palchikov replied:
I'll deduplicate in a follow-up PR.

Joerger commented on api/client/sessions.go:
```diff
@@ -0,0 +1,152 @@
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
```
`lib/services` can't be imported here because it brings a ton of dependencies into `api` that we don't want. I can help refactor WebSessions to solve this.

a-palchikov committed:
Address review comments

a-palchikov committed:
Only set HTTP fs on the web handler if the UI is enabled

a-palchikov committed:
Address review comments

a-palchikov committed:
Properly tear down web session test by releasing resources at the end.
Fix the web UI assets configuration by removing DisableUI and instead
use the presence of assets (HTTP file system) as an indicator that
the web UI has been enabled.

a-palchikov committed:
Decrease the expired session cache clean up threshold to 2m. Only log the expiration error message for errors other than not found

a-palchikov committed:
Add test for terminal disconnect when using two proxies in HA mode

a-palchikov committed:
Fix rebase issues to sync with master
